torch_xla/experimental/pjrt.py

```diff
@@ -0,0 +1,216 @@
+import concurrent.futures
+import functools
+import os
+import threading
+from typing import Callable, Dict, Optional, TypeVar
+
+import torch
+import torch_xla
+import torch_xla.core.xla_env_vars as xenv
+import torch_xla.core.xla_model as xm
+import torch_xla.utils.utils as xu
+
+_PJRT_ORDINALS = threading.local()
+
+R = TypeVar('R')
+FN = TypeVar('FN')
+
+
+def set_device_type(pjrt_device: str) -> None:
+  """Sets the current PjRt device type.
+
+  Must be run before using any XLA devices.
+
+  Args:
+    pjrt_device: 'TPU' or 'CPU'
+  """
+  os.environ[xenv.PJRT_DEVICE] = pjrt_device
+
+
+def device_type() -> Optional[str]:
+  """Returns the currrent PjRt device type."""
+  return xu.getenv_as(xenv.PJRT_DEVICE, str)
+
+
+def using_pjrt() -> bool:
+  """Returns whether this process is using PjRt runtime."""
+  return device_type() is not None
+
+
+def num_visible_tpu_chips(default: int = 4) -> int:
+  """Returns number of TPU chips visible to current process."""
+  visible_devices = xu.getenv_as(xenv.TPU_VISIBLE_DEVICES, str)
+
+  return len(visible_devices.split(',')) if visible_devices else default
+
+
+def configure_tpu_topology(rank: int, processes: int, base_port=8476) -> None:
```

Comment:
how is the single process PJRT example looks like? Would user needs to set any of below env vars?

Reply:
Yes. If the user sets `TPU_VISIBLE_DEVICES=[0 .. 3]` and `TPU_PROCESS_BOUNDS=1,1,1`, the other variables will be inferred.

Reply:
Yes, you can still just set `PJRT_DEVICE=TPU` and use one core, but the client will be initialized with 8 devices, preventing any other process from using them.